cpp/include/algorithms.hpp

```diff
@@ -1187,5 +1187,33 @@ extract_ego(raft::handle_t const &handle,
             vertex_t *source_vertex,
             vertex_t n_subgraphs,
             vertex_t radius);
+
+/**
+ * @brief returns random walks (RW) from starting sources, where each path is of given maximum
+ * length. Uniform distribution is assumed for the random engine.
+ *
+ * @tparam graph_t Type of graph.
+ * @tparam vertex_type Type of vertex identifiers. Needs to be an integral type.
+ * @tparam weight_type Type of edge weights. Needs to be a floating point type.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph Graph object to generate RW on.
+ * @param ptr_d_start Device pointer to set of starting vertex indices for the RW.
+ * @param num_paths = number(paths) == d_v_start.size().
+ * @param max_depth maximum length of RWs.
+ * @return std::tuple<device_vec_t<vertex_t>, device_vec_t<weight_t>,
+ * device_vec_t<index_t>> Triplet of coalesced RW paths, with corresponding edge weights for
+ * each, and coresponding path sizes. This is meant to minimize the number of DF's to be passed to
+ * the Python layer.
+ */
+template <typename graph_t, typename index_t>
+std::tuple<rmm::device_uvector<typename graph_t::vertex_type>,
+           rmm::device_uvector<typename graph_t::weight_type>,
+           rmm::device_uvector<index_t>>
+random_walks(raft::handle_t const &handle,
+             graph_t const &graph,
+             typename graph_t::vertex_type const *ptr_d_start,
+             index_t num_paths,
+             index_t max_depth);
```

Comment:
number of edges typically is an `edge_t` type, including in code that has already been there for a while. For example `number_of_edges`. Since `num_paths` can be very large (larger than `number_of_edges` it defaults to `edge_t` but can be customized).

Reply:
num_paths is only used to convert the raw pointer/size back into a vector in your implementation.  That should probably be a size_t.

Do we expect the max_depth to be a larger integer?  Would hardcoding that to an int32_t be sufficient?  The returned results is going to be O(num_paths * max_depth).  That will limit the value that you can pass in to max_depth anyway.

Reply:
(1.) `num_paths` can be higher than `num_vertices` because you can have repetitions (in fact, that's the typical usage of RW I've seen in the literature: very many paths). Hence, `O(num_paths * max_depth)` might need larger storage.
(2.) I've followed the model in `graph_view_t`, `graph_t`, and `graph_base_t` where sizes (`number_of_vertices`, `number_of_edges`) have parameterized types (`vertex_t`, `edge_t`, respectively). I'm not sure why it was okay there but not here.
(3.) Also, why back-n-forth type conversions if they can be avoided? They don't add performance...but they make code less readable (and maintainable).

Reply:
I wouldn't characterize my concern as not feeling like it's OK here.  I'm mainly challenging the need to create an extra template parameter that needs to be understood by the caller and maintainers of the code to understand if it truly adds value or if it could be avoided.  We had a solid set of reasoning behind `vertex_t` and `edge_t` being separate template parameters.  I wanted to understand the reasoning behind adding another template typename vice using a simple type or one of the existing template typename parameters.

I can see benefit to having index_t called out separately, although it's not clear to me (at the moment) that it will ever be anything but an int32_t.